### PR TITLE
Avoid creating unnecessary Sets and avoid resizing in removeBlocks

### PR DESCRIPTION
**Pull request by jiacheliu3**

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 1]
```diff
@@ -665,14 +665,15 @@ private WorkerInfo extractWorkerInfo(MasterWorkerInfo worker,
...
+        Set<Long> workerIds;
         try (LockResource r = lockBlock(blockId)) {
           Optional<BlockMeta> block = mBlockStore.getBlock(blockId);
           if (!block.isPresent()) {
             continue;
           }
-          for (BlockLocation loc : mBlockStore.getLocations(blockId)) {
+          List<BlockLocation> locations = mBlockStore.getLocations(blockId);
+          workerIds = new HashSet<>(locations.size());
+          for (BlockLocation loc : locations) {
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:676]: @dbw9580 Do you happen to know how I can do things like below?
```
new HashSet<>(mBlockStore.getLocations(blockId).stream().map(BlockLocation::getWorkerId))
```
Basically I want to use the sized constructor. 
```
list.stream.map(...).collect(Collectors.toSet)
```
This is actually using the non-sized constructor of HashSet and repeatedly doing set.addAll() which causes resizing.

Comment (dbw9580) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:676]: `.collect(Collectors.toCollection(() -> new HashSet<>(locations.size())))`

The readability is subtly inferior to a for loop, IMO.

Comment (dbw9580) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:676]: And does not using streams create even more intermediate, ephemeral objects behind the scenes, considering this piece of code is inside an outer for loop?

Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:676]: Ok i'll stick to for loop then. Thx :)

**Review by jenoudet — APPROVED** [review]

**Comment by jiacheliu3** [comment]
The failed test case is a know flake:
```
Error: 9.600 [ERROR] Failures: 
Error: 9.601 [ERROR] alluxio.client.fs.io.FileOutStreamIntegrationTest.cancelWrite[2]
Error: 9.603 [ERROR]   Run 1: FileOutStreamIntegrationTest.cancelWrite:248 expected:<0> but was:<1000>
Error: 9.603 [ERROR]   Run 2: FileOutStreamIntegrationTest.cancelWrite:248 expected:<0> but was:<2000>
```

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-2.7 please

**Pull request merged**